src/txHashBuilder.c

```diff
@@ -1047,11 +1077,93 @@ static void txHashBuilder_assertCanLeaveValidityIntervalStart(tx_hash_builder_t*
 	}
 }
 
-void txHashBuilder_finalize(tx_hash_builder_t* builder, uint8_t* outBuffer, size_t outSize)
+void txHashBuilder_enterMint(tx_hash_builder_t* builder)
 {
 	_TRACE("state = %d", builder->state);
 
 	txHashBuilder_assertCanLeaveValidityIntervalStart(builder);
+	{
+		// Enter mint
+		BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, TX_BODY_KEY_MINT);
+	}
+	builder->state = TX_HASH_BUILDER_IN_MINT;
+}
+
+void txHashBuilder_addMint_topLevelData(
+        tx_hash_builder_t* builder, uint16_t numAssetGroups
+)
+{
+	_TRACE("state = %u", builder->state);
+
+	ASSERT(builder->state == TX_HASH_BUILDER_IN_MINT);
+
+	builder->multiassetData.remainingAssetGroups = numAssetGroups;
+	// Map(numAssetGroups)[
+	//   { * policy_id => { * asset_name => uint } }
+	// ]
+	BUILDER_APPEND_CBOR(CBOR_TYPE_MAP, numAssetGroups);
+	if (0 != numAssetGroups) {
+		builder->state = TX_HASH_BUILDER_IN_MINT_ASSET_GROUP;
+	} else {
+		// TODO is this an error case?
```

Comment:
I don't think there is any point in serializing empty asset groups. Choose the way that is simpler --- just serialize it correctly without validation (might work out of the box with empty cbor arrays), or ASSERT num > 0 and avoid serializing altogether if it would add addtional complexity. (I guess we can have the assertion in both cases anyway.)